### ISSUE TITLE
Reproducible training

### ISSUE THREAD

**Issue by user799595** [UKPLab/sentence-transformers#3248]
Dear Sentence Transformers team,

Is it possible to get reproducible training runs with sentence transformers?

I'm having an issue that I get slightly different loss curves every time I run my train script.

I realize this might be an upstream issue with transformers/pytorch/Cuda.

What I've tried:
transformers.set_seed
seed in training_arguments
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
torch.backends.cudnn.enabled = False

Thank you!

**Comment by tomaarsen** [comment]
Hello!

Yes, it should be possible to get reproducible scripts. It should even happen out of the box.
I just ran this short script twice, and I got the same losses on both runs:
```python
from datasets import load_dataset
from sentence_transformers import (
    SentenceTransformer,
    SentenceTransformerTrainer,
    SentenceTransformerTrainingArguments,
)
from sentence_transformers.losses import MultipleNegativesRankingLoss
from sentence_transformers.training_args import BatchSamplers

# 1. Load a model to finetune
model = SentenceTransformer("microsoft/mpnet-base")

# 3. Load a dataset to finetune on
dataset = load_dataset("sentence-transformers/all-nli", "triplet")
train_dataset = dataset["train"].select(range(1_000))
eval_dataset = dataset["dev"]
test_dataset = dataset["test"]

# 4. Define a loss function
loss = MultipleNegativesRankingLoss(model)

# 5. (Optional) Specify training arguments
args = SentenceTransformerTrainingArguments(
    # Required parameter:
    output_dir="models/mpnet-base-all-nli-triplet",
    # Optional training parameters:
    num_train_epochs=1,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    learning_rate=2e-5,
    warmup_ratio=0.1,
    fp16=True,  # Set to False if you get an error that your GPU can't run on FP16
    bf16=False,  # Set to True if you have a GPU that supports BF16
    batch_sampler=BatchSamplers.NO_DUPLICATES,  # MultipleNegativesRankingLoss benefits from no duplicate samples in a batch
    # Optional tracking/debugging parameters:
    eval_strategy="no",
    save_strategy="no",
    save_total_limit=2,
    logging_steps=10,
    run_name="mpnet-base-all-nli-triplet",  # Will be used in W&B if `wandb` is installed
)

# 7. Create a trainer & train
trainer = SentenceTransformerTrainer(
    model=model,
    args=args,
    train_dataset=train_dataset,
    loss=loss,
)
trainer.train()

# 8. Save the trained model
model.save_pretrained("models/mpnet-base-all-nli-triplet/final")
```
Run 1:
```
{'loss': 3.2453, 'grad_norm': 16.65165138244629, 'learning_rate': 1.9642857142857145e-05, 'epoch': 0.16}
{'loss': 2.2494, 'grad_norm': 15.62674617767334, 'learning_rate': 1.642857142857143e-05, 'epoch': 0.32}                                                            
{'loss': 1.3796, 'grad_norm': 36.987300872802734, 'learning_rate': 1.3214285714285716e-05, 'epoch': 0.48}                                                          
{'loss': 0.6697, 'grad_norm': 8.747308731079102, 'learning_rate': 9.642857142857144e-06, 'epoch': 0.63}                                                            
{'loss': 0.466, 'grad_norm': 10.980976104736328, 'learning_rate': 6.071428571428571e-06, 'epoch': 0.79}                                                            
{'loss': 0.394, 'grad_norm': 50.226585388183594, 'learning_rate': 2.5e-06, 'epoch': 0.95}                                                                          
{'train_runtime': 19.3111, 'train_samples_per_second': 51.784, 'train_steps_per_second': 3.262, 'train_loss': 1.3594722331516327, 'epoch': 1.0}
```
Run 2:
```
{'loss': 3.2453, 'grad_norm': 16.65165138244629, 'learning_rate': 1.9642857142857145e-05, 'epoch': 0.16}
{'loss': 2.2494, 'grad_norm': 15.62674617767334, 'learning_rate': 1.642857142857143e-05, 'epoch': 0.32}                                                            
{'loss': 1.3796, 'grad_norm': 36.987300872802734, 'learning_rate': 1.3214285714285716e-05, 'epoch': 0.48}                                                          
{'loss': 0.6697, 'grad_norm': 8.747308731079102, 'learning_rate': 9.642857142857144e-06, 'epoch': 0.63}                                                            
{'loss': 0.466, 'grad_norm': 10.980976104736328, 'learning_rate': 6.071428571428571e-06, 'epoch': 0.79}                                                            
{'loss': 0.394, 'grad_norm': 50.226585388183594, 'learning_rate': 2.5e-06, 'epoch': 0.95}
{'train_runtime': 21.2168, 'train_samples_per_second': 47.133, 'train_steps_per_second': 2.969, 'train_loss': 1.3594722331516327, 'epoch': 1.0}
```
This is because ST is meant to be reproducible out of the gate, i.e. it sets a seed via the SentenceTransformerTrainingArguments behind the scenes, and tries to use that for everything that uses randomness.
However, I suspect that you're using 1) some option from ST that isn't correctly set up with this, or 2) some randomness outside of ST in your script.

Some questions:
* Do you set the `batch_sampler` or `multi_dataset_batch_sampler` in your `SentenceTransformerTrainingArguments`? If yes, to what? (FYI: I got the same losses by both `batch_sampler=BatchSamplers.NO_DUPLICATES` and not setting either of those arguments)
* Do you perhaps have some randomness in your data loading/preprocessing? 

I can't really think of anything else that might cause it. Perhaps an outdated `sentence_transformers` or `transformers` version?

- Tom Aarsen

**Comment by user799595** [comment]
Many thanks for the detailed response Tom. 

I did some more debugging and it turned out to be a Cuda issue. 

I was able to fix it by setting:
```
torch.use_deterministic_algorithms(True)
```

Interestingly, the non-determinism didn't arise using your script, until I plugged in our data, but I wasn't able to figure out what it was about our data that triggered it.

**Comment by user799595** [comment]
For future reference, there's actually a training argument that sets this `full_determinism`.
